src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -34,61 +34,97 @@ namespace NodeStore {
 
 constexpr std::uint32_t DatabaseShard::ledgersPerShardDefault;
 
-DatabaseShardImp::DatabaseShardImp(Application& app,
-    std::string const& name, Stoppable& parent, Scheduler& scheduler,
-    int readThreads, Section const& config, beast::Journal j)
+DatabaseShardImp::DatabaseShardImp(
+    Application& app,
+    std::string const& name,
+    Stoppable& parent,
+    Scheduler& scheduler,
+    int readThreads,
+    Section const& config,
+    beast::Journal j)
     : DatabaseShard(name, parent, scheduler, readThreads, config, j)
     , app_(app)
+    , ctx_(std::make_unique<nudb::context>())
     , config_(config)
     , dir_(get<std::string>(config, "path"))
     , backendName_(Manager::instance().find(
-        get<std::string>(config_, "type"))->getName())
+        get<std::string>(config_, "type", "nudb"))->getName())
     , maxDiskSpace_(get<std::uint64_t>(config, "max_size_gb") << 30)
     , ledgersPerShard_(get<std::uint32_t>(
         config, "ledgers_per_shard", ledgersPerShardDefault))
     , earliestShardIndex_(seqToShardIndex(earliestSeq()))
     , avgShardSz_(ledgersPerShard_ * (192 * 1024))
 {
-    if (ledgersPerShard_ == 0 || ledgersPerShard_ % 256 != 0)
-        Throw<std::runtime_error>(
-            "ledgers_per_shard must be a multiple of 256");
+    ctx_->start();
 }
 
 DatabaseShardImp::~DatabaseShardImp()
 {
     // Stop threads before data members are destroyed
     stopThreads();
+
+    std::lock_guard<std::mutex> lock(m_);
+    complete_.clear();
+    if (incomplete_)
+        incomplete_.reset();
+    try
+    {
+        ctx_.reset();
+    }
+    catch (std::exception const& e)
+    {
+        JLOG(j_.error()) <<
+            e.what();
+    }
 }
 
 bool
 DatabaseShardImp::init()
 {
     using namespace boost::filesystem;
+    using namespace boost::beast::detail;
+
     std::lock_guard<std::mutex> lock(m_);
     if (init_)
     {
-        assert(false);
         JLOG(j_.error()) <<
             "Already initialized";
         return false;
     }
-
-    // Find backend type and file handle requirement
-    try
+    if (ledgersPerShard_ == 0 || ledgersPerShard_ % 256 != 0)
     {
-        fdLimit_ = Manager::instance().make_Backend(
-            config_, scheduler_, j_)->fdlimit();
+        JLOG(j_.error()) <<
+            "ledgers_per_shard must be a multiple of 256";
+        return false;
     }
-    catch (std::exception const&)
+
+    // NuDB is the default and only supported permanent storage backend
```

Comment:
should this say `"nudb"` instead? (just to be consistent with how `"memory"` and `"none"` are written)

Reply:
No, since this is referring to the name. But the string literals should be "NuDB",  "Memory" and "none", thanks.